pkg/controller/master/rancher/namespace.go

```diff
@@ -0,0 +1,104 @@
+package rancher
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sschema "k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/harvester/harvester/pkg/settings"
+	"github.com/harvester/harvester/pkg/util"
+)
+
+func (h *Handler) onNamespaceRemoved(_ string, namespace *corev1.Namespace) (*corev1.Namespace, error) {
+	if namespace == nil {
+		return namespace, nil
+	}
+
+	value := settings.RancherCluster.Get()
+	if value == "" {
+		return namespace, nil
+	}
+
+	rancherClusterConfig, err := settings.DecodeConfig[settings.RancherClusterConfig](value)
+	if err != nil {
+		return namespace, err
+	}
+
+	if !rancherClusterConfig.RemoveUpstreamClusterWhenNamespaceIsDeleted {
+		return namespace, nil
+	}
+
+	secret, err := h.SecretCache.Get(util.HarvesterSystemNamespaceName, util.RancherClusterConfigSecretName)
```

Comment:
If the secret was created before the setting of `rancher-cluster-*`, and later the secret is removed;  the removing of normal namespace will be blocked here, even if the namespace has no relationship with Rancher Manager.